### ISSUE TITLE
QuickUMLS: Remove leveldb and its dependencies

### ISSUE THREAD

**Issue by burgersmoke** [medspacy/medspacy#142]
There are now multiple reasons that we are likely to remove leveldb entirely as an option and dependency in our fork of QuickUMLS:

1. Given issue #141 , match behavior for multi-word expressions does not work as expected
2. Currently leveldb requires specific post-installation steps using conda for Windows only which has required time to support (see https://github.com/medspacy/medspacy/blob/master/windows_and_quickumls.md)
3. The Python wrappers for leveldb have not seen a release since 2019 so support may be limited

Given these items and others, @turbosheep and I have discussed this and we are very likely to remove leveldb entirely from our fork of QuickUMLS.

**Comment by patrickthealba** [comment]
@jdposada mentioned some additional issues they were encountering with quickumls/medspacy today.  Sounds like they might not be directly related to leveldb. Commenting here to hopefully include them in any quickumls discussions.

**Comment by burgersmoke** [comment]
One other FYI on this.  I was hopeful that leveldb was the only other dependency for QuickUMLS which could not be installed with pip, and therefore had custom steps with conda.

However, QuickUMLS depends on simstring which depends on `libiconv` which cannot be installed trivially with pip.  So while we might break a dependency with leveldb, it looks like we will still have a Windows-only step of installing libiconv with conda.

**Comment by jdposada** [comment]
hi @burgersmoke and @patrickthealba 

Below are the things, I did to make things run. Each of of them may need a fix

- I created a Dockerfile to fix the libiconv issue on a Linux machine. The Dockerfileis below. Also, there is an issue importing spacy modules since the new spacy version changed how modules are imported. Even setting up this variable in the Dockerfile did not work. I actually needed to export it on the command line session. `LD_PRELOAD=/usr/local/lib/libiconv.so` not sure why. 

- The other issue was with a set of rules that were not found. I did a copy to the actual installation folder so the files were actually there.

- I changed line 156 on quickumls/core.py `self.nlp = spacy.load("en_core_web_sm")`. The params are not correct at the moment

- I was not able to use Medspacy Spacy pipeline. I used QuickUMLS spacy pipeline and then added medspacy as a module.
```python
nlp = spacy.load('en_core_web_sm')

@Language.factory("quickumls_component")
def create_pyrush(nlp, name):
    return SpacyQuickUMLS(nlp, '/workspaces/QuickUMLS/umls/qumls_db/')

nlp.add_pipe('quickumls_component')
nlp.add_pipe('medspacy_context')
nlp.select_pipes(disable=["attribute_ruler", "lemmatizer", "ner"])
print(nlp.pipe_names)
```

## Dockerfile

```bash
FROM jupyter/minimal-notebook:python-3.9.12

USER root
# [Optional] Uncomment this section to install additional OS packages.
RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
    && apt-get -y install python3.9-distutils python3-dev build-essential

RUN wget http://ftp.gnu.org/pub/gnu/libiconv/libiconv-1.17.tar.gz \
&& tar zxvf libiconv-1.17.tar.gz \
&& cd libiconv-1.17 \ 
&& ./configure --prefix=/usr/local \ 
&& make \ 
&& make install 

COPY requirements.txt* noop.txt /tmp/conda-tmp/
RUN if [ -f "/tmp/conda-tmp/requirements.txt" ]; then umask 0002 && /opt/conda/bin/conda env update -n base -f /tmp/conda-tmp/requirements.txt; fi \
    && rm -rf /tmp/conda-tmp
RUN conda --version

COPY requirements_pip.txt /home
RUN pip install -r /home/requirements_pip.txt

RUN git -C /tmp clone https://github.com/Georgetown-IR-Lab/simstring.git
RUN cd /tmp/simstring && python setup.py build_ext --inplace && python setup.py install

ENV DYLD_LIBRARY_PATH=/usr/local/lib
ENV LD_LIBRARY_PATH=/usr/local/lib

RUN git -C /tmp clone https://github.com/Georgetown-IR-Lab/QuickUMLS.git
RUN cd /tmp/QuickUMLS && python setup.py install

RUN export PATH="/usr/local/lib:$PATH"
ENV LD_PRELOAD=/usr/local/lib/libiconv.so

RUN cp medspacy/resources/rush_rules.tsv /opt/conda/lib/python3.9/site-packages/conf/
```

**Comment by burgersmoke** [comment]
As far as the initial problem listed in this ticket, that has now been resolved with medspacy version 1.0.0 which was released yesterday.  

@jdposada I believe that several of the issues that you mention here have been resolved with this release.  Some will not be addressed since they are simply how operating systems find libraries (i.e., the LD_LIBRARY_PATH item mentioned).

If there are still individual issues which you feel are outstanding, could you write individual tickets for these?

As a side note, I see that you are gathering QuickUMLS and its dependencies (simstring) from the original QuickUMLS repo (https://github.com/Georgetown-IR-Lab).  As of medspacy 1.0.0, we are maintaining our own forks of QuickUMLS and simsring.  If you identify issues with our own forks (as opposed to the Georgetown-IR-Lab), can you report them here to this medspacy group and we can try to discuss solutions?

Thank you so much for this information.
